File: data/code_pairs/fortran/370.f90
module test_module
  implicit none

  type, public :: test_p
    private
    class (test_p), pointer :: next => null()
  end type test_p

  type, public :: test
    class (test), pointer :: next => null()
  end type test

contains

  subroutine do_it(x)
    class (test_p), target :: x
    x%next => x
  end subroutine do_it

end module test_module

program test_program
  use test_module
  implicit none

  class(test_p), pointer :: x

  ! Setup
  allocate(x)
  call do_it(x)

  ! Test to see if x%next is associated with x (simple test)
  if (associated(x%next, x)) then
    print *, "Test passed: x%next is associated with x."
  else
    print *, "Test failed: x%next is not associated with x."
  endif

  ! Teardown
  deallocate(x)
end program test_program

File: data/code_pairs/cpp/370.cpp
#include <iostream>
#include <memory>

class TestModule {
public:
    class TestP;
    class Test;

    class TestP {
    private:
        std::shared_ptr<TestP> next;

    public:
        TestP() : next(nullptr) {}

        void setNext(const std::shared_ptr<TestP>& nextPtr) {
            this->next = nextPtr;
        }

        std::shared_ptr<TestP> getNext() const {
            return next;
        }
    };

    class Test {
    public:
        std::shared_ptr<Test> next;

        Test() : next(nullptr) {}
    };

    static void doIt(std::shared_ptr<TestP>& x) {
        x->setNext(x);
    }
};

int main() {
    std::shared_ptr<TestModule::TestP> x = std::make_shared<TestModule::TestP>();
    TestModule::doIt(x);

    if(x == x->getNext()) {
        std::cout << "Test passed: x->next is pointing to x." << std::endl;
    } else {
        std::cout << "Test failed: x->next is not pointing to x." << std::endl;
    }

    return 0;
}